server/balancer.go

```diff
@@ -0,0 +1,316 @@
+
+// Do implements Operator.Do interface.
+func (bo *BalanceOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	// TODO: optimize it to return next operator response directly.
+	if bo.Check(req) {
+		return true, nil
+	}
+
+	finished, res := bo.ops[bo.index].Do(req)
+	if !finished {
+		return false, res
+	}
+
+	bo.index++
+	return false, nil
+}
+
+// ChangePeerOperator is used to do peer change.
+type ChangePeerOperator struct {
+	changePeer *pdpb.ChangePeer
+}
+
+func newChangePeerOperator(changePeer *pdpb.ChangePeer) *ChangePeerOperator {
+	return &ChangePeerOperator{
+		changePeer: changePeer,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (co *ChangePeerOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	region := req.GetRegion()
+	if region == nil {
+		return false
+	}
+
+	leaderPeer := req.GetLeader()
+	if leaderPeer == nil {
+		return false
+	}
+
+	if co.changePeer.GetChangeType() == raftpb.ConfChangeType_AddNode {
+		if containPeer(region, leaderPeer) {
+			return true
+		}
+	} else if co.changePeer.GetChangeType() == raftpb.ConfChangeType_RemoveNode {
+		if !containPeer(region, leaderPeer) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// Do implements Operator.Do interface.
+func (co *ChangePeerOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	if co.Check(req) {
+		return true, nil
+	}
+
+	res := &pdpb.RegionHeartbeatResponse{
+		ChangePeer: co.changePeer,
+	}
+	return false, res
+}
+
+// TransferLeaderOperator is used to do leader transfer.
+type TransferLeaderOperator struct {
+	oldLeader *metapb.Peer
+	newLeader *metapb.Peer
+}
+
+func newTransferLeaderOperator(oldLeader, newLeader *metapb.Peer) *TransferLeaderOperator {
+	return &TransferLeaderOperator{
+		oldLeader: oldLeader,
+		newLeader: newLeader,
+	}
+}
+
+// Check implements Operator.Check interface.
+func (lto *TransferLeaderOperator) Check(req *pdpb.RegionHeartbeatRequest) bool {
+	leader := req.GetLeader()
+	if leader == nil {
+		return false
+	}
+
+	if leader.GetId() == lto.newLeader.GetId() {
+		return true
+	}
+
+	return false
+}
+
+// Do implements Operator.Doop interface.
+func (lto *TransferLeaderOperator) Do(req *pdpb.RegionHeartbeatRequest) (bool, *pdpb.RegionHeartbeatResponse) {
+	if lto.Check(req) {
+		return true, nil
+	}
+
+	// TODO: call admin command, maybe we can send it to a channel.
+	/*
+		res := &raft_cmdpb.AdminRequest{
+			CmdType: raft_cmdpb.AdminCmdType_TransferLeader.Enum(),
+			TransferLeader: &raft_cmdpb.TransferLeaderRequest{
+				Peer: lto.newLeader,
+			},
+		}
+	*/
+
+	return false, nil
+}
+
+// Balancer is an interface to select store regions for auto-balance.
+type Balancer interface {
+	Balance(cluster *ClusterInfo, c *raftCluster) (*BalanceOperator, error)
+}
+
+const (
+	// If the used ratio of one storage is greater than this value,
+	// it will never be used as a selected target and it should be rebalanced.
+	maxCapacityUsedRatio = 0.4
```

Comment:
I suggest that we set two thresholds. The first one is soft threshold which is lower. The second one is hard threshold which is the same as  maxCapacityUsedRatio. If the used ratio is lower than soft threshold, we should ignore it.
The operator on region with used ratio lower than hard threshold should has low priority. How about we implement this strategy in the future PR?